OpenRA.Mods.Common/Traits/World/PaletteFromGimpFile.cs

```diff
@@ -0,0 +1,110 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2017 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.IO;
+using OpenRA.Graphics;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	[Desc("Load a GIMP .gpl Palette file.")]
+	class PaletteFromGimpFileInfo : ITraitInfo
+	{
+		[FieldLoader.Require, PaletteDefinition]
+		[Desc("Palette name used internally.")]
+		public readonly string Name = null;
+
+		[Desc("If defined, load the palette only for this tileset.")]
+		public readonly string Tileset = null;
+
+		[FieldLoader.Require]
+		[Desc("Name of the file to load.")]
+		public readonly string Filename = null;
+
+		[Desc("Map listed indices to shadow. Ignores previous color.")]
+		public readonly int[] ShadowIndex = { };
+
+		public readonly bool AllowModifiers = true;
+
+		public object Create(ActorInitializer init) { return new PaletteFromGimpFile(init.World, this); }
+	}
+
+	class PaletteFromGimpFile : ILoadsPalettes, IProvidesAssetBrowserPalettes
+	{
+		readonly World world;
+		readonly PaletteFromGimpFileInfo info;
+
+		public PaletteFromGimpFile(World world, PaletteFromGimpFileInfo info)
+		{
+			this.world = world;
+			this.info = info;
+		}
+
+		public void LoadPalettes(WorldRenderer wr)
+		{
+			var colors = new uint[Palette.Size];
+			using (var s = world.Map.Open(info.Filename))
+			{
+				var lines = s.ReadAllLines().GetEnumerator();
+
+				lines.MoveNext();
+
+				if (lines.Current != "GIMP Palette")
+					throw new InvalidDataException("File {0} is not a valid GIMP palette.".F(info.Filename));
+
+				int r, g, b = 0;
+				var i = 0;
+
+				while (lines.MoveNext() && i < Palette.Size)
+				{
+					// Skip until first color. Ignore # comments, Name/Columns and blank lines.
+					if (string.IsNullOrEmpty(lines.Current) || !char.IsDigit(lines.Current.Trim()[0]))
+						continue;
+
+					var rgbPlusColorName = lines.Current.Split((char[])null, StringSplitOptions.RemoveEmptyEntries);
+					if (rgbPlusColorName.Length < 3)
+						throw new InvalidDataException("Invalid RGB triplet: {0}".F(rgbPlusColorName));
+
+					if (!int.TryParse(rgbPlusColorName[0], out r))
+						throw new InvalidDataException("Invalid R value: {0}".F(rgbPlusColorName[0]));
+
+					if (!int.TryParse(rgbPlusColorName[1], out g))
+						throw new InvalidDataException("Invalid G value: {0}".F(rgbPlusColorName[1]));
+
+					if (!int.TryParse(rgbPlusColorName[2], out b))
```

Comment:
Don't we actually want `byte.TryParse` for all of these? `Color.FromArgb` accepts ints for ease of use, but actually only accepts values of 0-255.